### PR TITLE
Supports Direct Private Key Authentication

### PR DESCRIPTION
**Pull request by mariotaku**
This PR added function to create SshKey from file or key content, and function to authenticate directly using this key.

### CODE REVIEW

File: libssh-rs/src/lib.rs [hunk 1]
```diff
@@ -1122,6 +1150,36 @@ impl SshKey {
...
+    pub fn from_privkey_file(filename: &str, passphrase:Option<&str>) -> SshResult<SshKey> {
+        let filename = CString::new(filename)
+            .map_err(|e| Error::Fatal(format!("Failed to process filename: {:?}", e)))?;
+        let passphrase = opt_str_to_cstring(passphrase);
+        unsafe {
+            let mut key = sys::ssh_key_new();
+            if sys::ssh_pki_import_privkey_file(filename.as_ptr(), opt_cstring_to_cstr(&passphrase),
+                                                  None, null_mut(), &mut key) != sys::SSH_OK as i32 {
+                sys::ssh_key_free(key);
+                return Err(Error::Fatal(format!("Failed to parse ssh key")));
```
Comment (wez) [inline libssh-rs/src/lib.rs:1178]:
```suggestion
                return Err(Error::Fatal(format!("Failed to parse ssh key from file '{filename}'")));
```

Can we get a better error reason from libssh for this case?

File: libssh-rs/src/lib.rs [hunk 2]
```diff
@@ -1122,6 +1150,36 @@ impl SshKey {
...
+            }
+            return Ok(SshKey { key });
+        }
+    }
+
+    pub fn from_privkey_file(filename: &str, passphrase:Option<&str>) -> SshResult<SshKey> {
+        let filename = CString::new(filename)
+            .map_err(|e| Error::Fatal(format!("Failed to process filename: {:?}", e)))?;
```
Comment (wez) [inline libssh-rs/src/lib.rs:1171]:
```suggestion
        let filename_cstr = CString::new(filename)
            .map_err(|e| Error::Fatal(format!("Could not make CString from filename '{filename}': {err:#}")))?;
```

File: libssh-rs/src/lib.rs [hunk 3]
```diff
@@ -1122,6 +1150,36 @@ impl SshKey {
...
+        let passphrase = opt_str_to_cstring(passphrase);
+        unsafe {
+            let mut key = sys::ssh_key_new();
+            if sys::ssh_pki_import_privkey_file(filename.as_ptr(), opt_cstring_to_cstr(&passphrase),
```
Comment (wez) [inline libssh-rs/src/lib.rs:1175]:
```suggestion
            if sys::ssh_pki_import_privkey_file(filename_cstr.as_ptr(), opt_cstring_to_cstr(&passphrase),
```

**Review by wez — COMMENTED** [review]
Thanks!
I'd like to have a bit more context in the error messages; some suggestions inline.
Please also `cargo fmt`!

**Comment by mariotaku** [comment]
Sorry for missing that, all done!

There are several features I'd like to add in the future by the way, such as scp. Are you okay with that too?

**Comment by wez** [comment]
> Sorry for missing that, all done!

Did you forget to push? :)

> There are several features I'd like to add in the future by the way, such as scp. Are you okay with that too?

I'm open to it!

Thanks!

**Comment by mariotaku** [comment]
Am I missing something? I ran `cargo fmt` and it gave me the last commit.

**Comment by wez** [comment]
Thanks!

**Pull request merged**